## Load and Preprocess the Image: 
1. The image is loaded and resized to standard dimensions (1024x1024).
2.	Generate SHA-256 Hash of the Image: The image is converted to bytes, and a SHA-256 hash is generated.
3.	Generate 128 Unique 128-bit Segments: The image hash is concatenated with an incrementing counter and hashed again to produce unique 128-bit segments.
4.	Calculate Pixel Coordinates: Each 128-bit segment is split into two 64-bit parts, and modulo operation is used to generate coordinates.
5.	Extract RGB Values: The RGB values of the selected pixels are extracted.
6.	Concatenate RGB Values into Binary String: The RGB values are concatenated to form a binary string.
7.	Convert Binary String to Mnemonic Seed Phrase: The binary string is converted to a mnemonic seed phrase using BIP-39 standards.
8.	Derive Keys from Mnemonic Seed: The mnemonic seed is converted to a binary seed using PBKDF2, and then private and public keys are derived.

### Install the libraries

In [1]:
!pip3 install --upgrade pip
!pip3 install mnemonic
!pip3 install pillow
!pip3 install pbkdf2

---

### Import all the libraries

In [2]:
import hashlib
import hmac
import binascii
from PIL import Image
from pbkdf2 import PBKDF2
import requests
import tkinter as tk
from tkinter import filedialog

### Import an Image

In [3]:
# Function to select an image file
def select_image():
    root = tk.Tk()
    root.withdraw()  # Hide the root window
    filename = filedialog.askopenfilename(title="Select an Image", filetypes=[("Image files", "*.png *.jpg *.jpeg *.bmp")])
    root.destroy()  # Destroy the root window
    return filename

# Function to load and preprocess the image
def load_image(image_path):
    image = Image.open(image_path)
    image = image.resize((1024, 1024))  # Resize to standard dimensions
    return image

# Function to save the resized image
def save_resized_image(image, save_path):
    image.save(save_path)

### Import the BIP-39 wordlist

In [4]:
# Function to download and save the BIP-39 wordlist to a file
def download_wordlist(url='https://raw.githubusercontent.com/bitcoin/bips/master/bip-0039/english.txt', filename='bip39_wordlist.txt'):
    response = requests.get(url)
    response.raise_for_status()  # Ensure we notice bad responses
    with open(filename, 'w') as file:
        file.write(response.text)

# Function to load the BIP-39 wordlist from a file
def load_wordlist(filename='bip39_wordlist.txt'):
    with open(filename, 'r') as file:
        wordlist = [line.strip() for line in file.readlines()]
    return wordlist

### Processing

In [5]:
# Function to generate SHA-256 hash of the image
def image_hash(image):
    image_data = image.tobytes()
    hash_object = hashlib.sha256(image_data)
    return hash_object.hexdigest()

# Function to generate 128 unique 128-bit segments
def generate_segments(image_hash):
    segments = []
    for i in range(128):
        hash_input = image_hash + str(i)
        hash_output = hashlib.sha256(hash_input.encode()).hexdigest()
        segment = hash_output[:32]  # First 128 bits (32 hex characters)
        segments.append(segment)
    return segments

# Function to calculate pixel coordinates from 128-bit segment
def segment_to_coordinates(segment):
    x = int(segment[:16], 16) % 1024
    y = int(segment[16:], 16) % 1024
    return (x, y)

# Function to extract RGB values from selected pixels
def extract_rgb_values(image, coordinates):
    pixel_data = []
    for coord in coordinates:
        pixel = image.getpixel(coord)
        pixel_data.append(pixel)
    return pixel_data

# Function to concatenate RGB values into binary string
def concatenate_pixel_data(pixel_data):
    binary_string = ''
    for pixel in pixel_data:
        binary_string += '{:08b}{:08b}{:08b}'.format(pixel[0], pixel[1], pixel[2])
    return binary_string

# Updated function to convert binary string to mnemonic seed phrase
def binary_to_mnemonic(binary_string, wordlist):
    # Use only the first 256 bits (32 bytes) for practical purposes
    binary_segment = binary_string[:256]
    
    # Convert binary_segment to hexadecimal and calculate the SHA-256 hash
    hex_segment = hex(int(binary_segment, 2))[2:].zfill(64)
    checksum = hashlib.sha256(binascii.unhexlify(hex_segment)).hexdigest()[:1]  # First 4 bits of the hash
    
    # Convert the checksum to binary and append it to the binary segment
    checksum_binary = bin(int(checksum, 16))[2:].zfill(4)
    entropy_with_checksum = binary_segment + checksum_binary
    
    # Convert the entropy_with_checksum to words
    words = []
    for i in range(0, len(entropy_with_checksum), 11):
        index = int(entropy_with_checksum[i:i+11], 2)
        words.append(wordlist[index])
    return ' '.join(words)

# Function to derive keys from mnemonic seed
def mnemonic_to_seed(mnemonic, passphrase=''):
    salt = 'mnemonic' + passphrase
    seed = PBKDF2(mnemonic, salt, iterations=2048, macmodule=hmac, digestmodule=hashlib.sha512).read(64)
    return seed

# Function to derive private and public keys using ECDSA
def derive_keys(seed):
    private_key = hashlib.sha256(seed[:32]).hexdigest()
    # Public key derivation would normally involve elliptic curve multiplication,
    # for simplicity, we demonstrate it here with a hash
    public_key = hashlib.sha256(private_key.encode()).hexdigest()
    return private_key, public_key

### Execution

In [6]:
# Example usage
if __name__ == "__main__":
    # Ensure the wordlist is downloaded
    download_wordlist()
    
    # Load the wordlist from the file
    wordlist = load_wordlist()
    
    # Let the user select an image
    image_path = select_image()
    if not image_path:
        print("No image selected. Exiting...")
        exit()

    image = load_image(image_path)

    # Save the resized image
    resized_image_path = "resized_image.png"  # You can choose a different path or filename
    save_resized_image(image, resized_image_path)
    print(f"Resized image saved as {resized_image_path}")
    
    img_hash = image_hash(image)
    segments = generate_segments(img_hash)
    coordinates = [segment_to_coordinates(segment) for segment in segments]
    pixel_data = extract_rgb_values(image, coordinates)
    binary_string = concatenate_pixel_data(pixel_data)
    mnemonic_seed = binary_to_mnemonic(binary_string, wordlist)
    seed = mnemonic_to_seed(mnemonic_seed)
    private_key, public_key = derive_keys(seed)
    
    print("Image:", image_path)
    print("Mnemonic Seed Phrase:", mnemonic_seed)
    print("Private Key:", private_key)
    print("Public Key:", public_key)

2024-07-06 15:50:14.605 Python[3697:8486052] WARNING: Secure coding is not enabled for restorable state! Enable secure coding by implementing NSApplicationDelegate.applicationSupportsSecureRestorableState: and returning YES.


Resized image saved as resized_image.png
Image: /Users/antonpiskunov/Pictures/Jackson Photos/IMG_6451.JPG
Mnemonic Seed Phrase: embark present make catch way seat change clap earth road mushroom champion speak own motor song grab hill bunker bring lift mention motor achieve
Private Key: ab3cf33a9eae186b2942da1015fae575f5823c7821093c891989dc7ed6f8b711
Public Key: 1eabdcfe0b5e5307472d6acb10baf68e8150ff6bda688d07dce06d133075ab32


In [2]:
!brew install gh

Running `brew update --auto-update`...
==> Downloading https://ghcr.io/v2/homebrew/portable-ruby/portable-ruby/blobs/sha256:a5ef040e054444a0eb2cbcc1032fed14702dfbe2e55b25e609f3ce643f23c4ee
######################################################################### 100.0%
==> Pouring portable-ruby-3.3.3.el_capitan.bottle.tar.gz
==> Auto-updated Homebrew!
Updated 2 taps (homebrew/core and homebrew/cask).
==> New Formulae
actions-batch              haproxy@2.8                phodav
age-plugin-se              iamb                       pkl
ansible-creator            ingress2gateway            pnpm@8
ansible@9                  inlyne                     podlet
apache-flink-cdc           jnv                        policy_sentry
appwrite                   jsontoolkit                porter
asak                       jtbl                       poutine
autotrace                  kconf                      prjtrellis
awsdac                     kubecolor                  promptfoo
basti             

# Evaluation

1. Entropy Calculation: The calculate_entropy function calculates the entropy based on the number of unique colors found in randomly selected pixels.
2.	Collision Resistance: The check_collision_resistance function checks if any hashes collide among a set of images.
3.	Brute-force Resistance: The calculate_brute_force_resistance function calculates the bits of security based on the number of possible pixel combinations.
4.	Statistical Randomness: The check_statistical_randomness function uses the chi-square test to verify the randomness of pixel values.

## Cryptographic Evaluation of Image-Based and Traditional Seed Phrase Methods

### Objective: To compare the cryptographic robustness and strength of private keys generated using image-based methods versus traditional BIP-39 seed phrases.

### Cryptographic Logic

1.	Entropy:
* Definition: Entropy measures the unpredictability or randomness in a system. Higher entropy indicates more randomness and better security.
* Image-Based Method: Entropy is derived from the unique pixel values in the image. Larger images with more unique colors provide higher entropy.
* Traditional BIP-39 Seed Phrase: Each word in the BIP-39 word list represents 11 bits of entropy. A 12-word seed phrase provides 128 bits of entropy.
2.	Collision Resistance:
* Definition: Collision resistance ensures that two different inputs do not produce the same hash output.
* Image-Based Method: The hash of the entire image data should be unique for different images.
* Traditional BIP-39 Seed Phrase: Each seed phrase should produce a unique hash.
3.	Brute-force Resistance:
* Definition: Brute-force resistance measures the difficulty of guessing a secret through exhaustive search.
* Image-Based Method: The large number of possible pixel combinations provides high brute-force resistance.
* Traditional BIP-39 Seed Phrase: The limited set of word combinations provides lower brute-force resistance compared to image-based methods.
4.	Statistical Randomness:
* Definition: Statistical randomness tests ensure that the output does not exhibit any predictable patterns.
* Image-Based Method: The chi-square test on pixel values helps verify the randomness of the image data.
* Traditional BIP-39 Seed Phrase: Deterministic in nature, hence this test is not applicable.

### Implementation

Steps:

1.	Generate Private Keys: Derive private keys from the seed phrases generated by both methods.
2.	Evaluate Cryptographic Strength: Compare entropy, collision resistance, brute-force resistance, and statistical randomness.

### Import libraries

In [7]:
import hashlib
import os
import random
import numpy as np
from PIL import Image
from collections import defaultdict
from scipy.stats import chisquare
import requests
import math
import secrets
import mnemonic
from concurrent.futures import ThreadPoolExecutor

### Image download

In [8]:
# URLs of the test images
image_urls = [
    "https://unsplash.com/photos/KMn4VEeEPR8/download?force=true&w=1024",
    "https://unsplash.com/photos/K_JPLwJhKdk/download?force=true&w=1024",
   # https://unsplash.com/photos/a-boat-that-is-sitting-on-the-ground-aNY0nGDcNtM
    "https://unsplash.com/photos/aNY0nGDcNtM/download?force=true&w=1024"
]

# Corresponding local file names for the downloaded images
image_filenames = ["image1.jpg", "image2.jpg", "image3.jpg"]

# Function to download images
def download_images(urls, filenames):
    for url, filename in zip(urls, filenames):
        response = requests.get(url)
        response.raise_for_status()  # Ensure we notice bad responses
        with open(filename, 'wb') as file:
            file.write(response.content)
        print(f"Downloaded {filename}")


In [9]:
# Function to calculate entropy for image-based method
def calculate_entropy(image_path, num_pixels=128):  # Increased number of pixels for higher accuracy
    image = Image.open(image_path).resize((1024, 1024))  # Use original 1024x1024 size
    img_data = np.array(image)
    unique_colors = set()
    for _ in range(num_pixels):
        x, y = random.randint(0, 1023), random.randint(0, 1023)
        unique_colors.add(tuple(img_data[x, y]))
    entropy = len(unique_colors) * 24
    return entropy

# Function to check collision resistance for image-based method
def check_collision_resistance(image_paths):
    hashes = defaultdict(int)
    for path in image_paths:
        image = Image.open(path).resize((1024, 1024))  # Use original 1024x1024 size
        img_hash = hashlib.sha256(image.tobytes()).hexdigest()
        hashes[img_hash] += 1
        if hashes[img_hash] > 1:
            return False, img_hash
    return True, None

# Function to calculate brute-force resistance for image-based method
def calculate_brute_force_resistance(num_pixels=128):  # Increased number of pixels for higher accuracy
    total_pixels = 1024 * 1024  # Total pixels in original 1024x1024 image
    combinations = math.comb(total_pixels, num_pixels)
    brute_force_bits = math.log2(combinations)
    return brute_force_bits

# Function to check statistical randomness for image-based method
def check_statistical_randomness(image_path, num_samples=1000):  # Increased number of samples for higher accuracy
    image = Image.open(image_path).resize((1024, 1024))  # Use original 1024x1024 size
    img_data = np.array(image)
    pixel_values = []
    for _ in range(num_samples):
        x, y = random.randint(0, 1023), random.randint(0, 1023)
        pixel_values.append(img_data[x, y][0])  # Using only the Red channel for simplicity
    observed_freq = np.bincount(pixel_values, minlength=256)
    expected_freq = np.ones(256) * num_samples / 256
    chi2_stat, p_value = chisquare(observed_freq, expected_freq)
    return chi2_stat, p_value

# Function to generate a traditional BIP-39 seed phrase
def generate_seed_phrase():
    mnemo = mnemonic.Mnemonic("english")
    entropy = secrets.token_bytes(16)  # 128 bits of entropy
    seed_phrase = mnemo.to_mnemonic(entropy)
    return seed_phrase

# Function to calculate entropy for traditional seed phrase method
def calculate_entropy_seed_phrase(seed_phrase):
    return 132

# Function to check collision resistance for traditional seed phrase method
def check_collision_resistance_seed_phrase(seed_phrases):
    hashes = defaultdict(int)
    for phrase in seed_phrases:
        phrase_hash = hashlib.sha256(phrase.encode()).hexdigest()
        hashes[phrase_hash] += 1
        if hashes[phrase_hash] > 1:
            return False, phrase_hash
    return True, None

# Function to calculate brute-force resistance for traditional seed phrase method
def calculate_brute_force_resistance_seed_phrase():
    brute_force_bits = 12 * math.log2(2048)
    return brute_force_bits

# Function to run all analyses for an image
def analyze_image(image_path):
    results = {}
    results['entropy'] = calculate_entropy(image_path)
    collision_free, colliding_hash = check_collision_resistance([image_path])
    results['collision_free'] = collision_free
    results['colliding_hash'] = colliding_hash
    results['brute_force_bits'] = calculate_brute_force_resistance()
    chi2_stat, p_value = check_statistical_randomness(image_path)
    results['chi2_stat'] = chi2_stat
    results['p_value'] = p_value
    return results

# Function to run all analyses for a seed phrase
def analyze_seed_phrase(seed_phrase):
    results = {}
    results['entropy'] = calculate_entropy_seed_phrase(seed_phrase)
    collision_free, colliding_hash = check_collision_resistance_seed_phrase([seed_phrase])
    results['collision_free'] = collision_free
    results['colliding_hash'] = colliding_hash
    results['brute_force_bits'] = calculate_brute_force_resistance_seed_phrase()
    return results

### Implementation

In [10]:
# Example usage
if __name__ == "__main__":
    # Download the test images
    download_images(image_urls, image_filenames)
    
    # Paths to the downloaded test images
    image_paths = image_filenames

    # Analyze images in parallel
    with ThreadPoolExecutor() as executor:
        image_results = list(executor.map(analyze_image, image_paths))

    for i, image_path in enumerate(image_paths):
        results = image_results[i]
        print(f"Results for {image_path} (image-based):")
        print(f"  Entropy: {results['entropy']} bits")
        print(f"  Collision Free: {results['collision_free']}")
        if not results['collision_free']:
            print(f"  Colliding Hash: {results['colliding_hash']}")
        print(f"  Brute-force resistance: {results['brute_force_bits']} bits")
        print(f"  Chi-square statistic: {results['chi2_stat']}, p-value: {results['p_value']}")

    # Generate and analyze traditional seed phrases
    seed_phrases = [generate_seed_phrase() for _ in range(3)]
    with ThreadPoolExecutor() as executor:
        seed_phrase_results = list(executor.map(analyze_seed_phrase, seed_phrases))

    for i, seed_phrase in enumerate(seed_phrases):
        results = seed_phrase_results[i]
        print(f"Results for Seed Phrase {i+1} (traditional):")
        print(f"  Seed Phrase: {seed_phrase}")
        print(f"  Entropy: {results['entropy']} bits")
        print(f"  Collision Free: {results['collision_free']}")
        if not results['collision_free']:
            print(f"  Colliding Hash: {results['colliding_hash']}")
        print(f"  Brute-force resistance: {results['brute_force_bits']} bits")

Downloaded image1.jpg
Downloaded image2.jpg
Downloaded image3.jpg
Results for image1.jpg (image-based):
  Entropy: 3048 bits
  Collision Free: True
  Brute-force resistance: 1843.8270944632009 bits
  Chi-square statistic: 1537.984, p-value: 5.641946164598493e-182
Results for image2.jpg (image-based):
  Entropy: 2904 bits
  Collision Free: True
  Brute-force resistance: 1843.8270944632009 bits
  Chi-square statistic: 4997.568, p-value: 0.0
Results for image3.jpg (image-based):
  Entropy: 2976 bits
  Collision Free: True
  Brute-force resistance: 1843.8270944632009 bits
  Chi-square statistic: 650.1759999999999, p-value: 2.345593322407705e-36
Results for Seed Phrase 1 (traditional):
  Seed Phrase: umbrella custom prison boss fetch spring extend great obvious prepare analyst pelican
  Entropy: 132 bits
  Collision Free: True
  Brute-force resistance: 132.0 bits
Results for Seed Phrase 2 (traditional):
  Seed Phrase: energy weapon spray body excite rival angry twelve oppose eye absent spar

## Overview:

The comparison involves evaluating the cryptographic robustness and security of private keys generated using image-based methods versus traditional BIP-39 seed phrases. The metrics used include entropy, collision resistance, brute-force resistance, and statistical randomness.

### Image-Based Method Results

Image 1:

•	Entropy: 3048 bits
•	Collision Free: True
•	Brute-force resistance: 1843.83 bits
•	Chi-square statistic: 1537.98, p-value: 5.64e-182

Image 2:

•	Entropy: 2904 bits
•	Collision Free: True
•	Brute-force resistance: 1843.83 bits
•	Chi-square statistic: 4997.57, p-value: 0.0

Image 3:

•	Entropy: 2976 bits
•	Collision Free: True
•	Brute-force resistance: 1843.83 bits
•	Chi-square statistic: 650.18, p-value: 2.35e-36

Traditional Seed Phrase Method Results

Seed Phrase 1:

•	Seed Phrase: umbrella custom prison boss fetch spring extend great obvious prepare analyst pelican
•	Entropy: 132 bits
•	Collision Free: True
•	Brute-force resistance: 132 bits

Seed Phrase 2:

•	Seed Phrase: energy weapon spray body excite rival angry twelve oppose eye absent spare
•	Entropy: 132 bits
•	Collision Free: True
•	Brute-force resistance: 132 bits

Seed Phrase 3:

•	Seed Phrase: consider tiger grass impose fitness sock scan giant fork donkey erosion double
•	Entropy: 132 bits
•	Collision Free: True
•	Brute-force resistance: 132 bits

### Detailed Analysis

#### Entropy

Definition: Entropy measures the unpredictability or randomness in a system. Higher entropy indicates more randomness and better security.

•	Image-Based Method: The entropy values for the images are significantly higher (3048, 2904, and 2976 bits) compared to the traditional method. This high entropy is due to the unique pixel values in the images, providing a vast amount of randomness.
•	Traditional Method: The entropy is fixed at 132 bits, standard for a 12-word BIP-39 seed phrase. While secure for practical purposes, it is significantly lower than the image-based method.

#### Collision Resistance

Definition: Collision resistance ensures that two different inputs do not produce the same hash output.

•	Image-Based Method: All images are collision-free, meaning each image generates a unique hash and consequently a unique seed phrase. This demonstrates good collision resistance.
•	Traditional Method: All seed phrases are collision-free, ensuring that each phrase generates a unique hash. This is expected given the design of BIP-39.

#### Brute-force Resistance

Definition: Brute-force resistance measures the difficulty of guessing a secret through exhaustive search.

•	Image-Based Method: The brute-force resistance is approximately 1843.83 bits, indicating an extremely large search space for potential brute-force attacks. This is due to the vast number of possible pixel combinations in a 1024x1024 image.
•	Traditional Method: The brute-force resistance is 132 bits, consistent with the entropy. While still secure, it is much lower than the image-based method.

#### Statistical Randomness

Definition: Statistical randomness tests ensure that the output does not exhibit any predictable patterns.

•	Image-Based Method: The chi-square test results indicate some deviation from expected randomness. High chi-square values and extremely low p-values (close to 0) suggest that the pixel values may not be perfectly random, but the vast amount of data (large images) compensates for it in terms of security.
•	Traditional Method: Not applicable, as BIP-39 seed phrases are deterministic and do not require a randomness test.

### Comparison and Conclusion

#### Image-Based Method

Advantages:

•	High Entropy: Provides significantly higher entropy compared to traditional methods, indicating better security.
•	High Brute-force Resistance: Extremely resistant to brute-force attacks due to the large number of possible combinations.
•	Unique Hashes: Demonstrates good collision resistance with unique hashes for each image.

Disadvantages:

•	Statistical Randomness: Chi-square test results suggest deviations from perfect randomness, but the high entropy compensates for this.
•	Practicality: Requires storing and processing large images, which can be cumbersome compared to text-based seed phrases.

#### Traditional Seed Phrase Method (BIP-39)

Advantages:

•	Standardized: Widely used and accepted in the cryptocurrency community.
•	Simplicity: Easy to generate, store, and use.
•	Secure: 132 bits of entropy is considered secure for most practical purposes.

Disadvantages:

•	Lower Entropy: Lower than the image-based method, indicating potentially less randomness.
•	Lower Brute-force Resistance: Easier to brute-force compared to the image-based method, though still extremely difficult in practice.


---